sound/soc/sof/ipc4-pcm.c

```diff
@@ -575,17 +577,48 @@ static void ipc4_ssp_dai_config_pcm_params_match(struct snd_sof_dev *sdev, const
 	if (!dai_link_found)
 		return;
 
+	/*
+	 * Find the first best matching hardware config:
+	 * rate + format + channels are matching
+	 * rate + format are matching
+	 * rate matching
+	 */
 	for (i = 0; i < slink->num_hw_configs; i++) {
 		struct snd_soc_tplg_hw_config *hw_config = &slink->hw_configs[i];
 
-		if (params_rate(params) == le32_to_cpu(hw_config->fsync_rate)) {
-			/* set current config for all DAI's with matching name */
-			list_for_each_entry(dai, &sdev->dai_list, list)
-				if (!strcmp(slink->link->name, dai->name))
-					dai->current_config = le32_to_cpu(hw_config->id);
+		if (params_rate(params) == le32_to_cpu(hw_config->fsync_rate) &&
+		    params_width(params) == le32_to_cpu(hw_config->tdm_slot_width) &&
+		    params_channels(params) == le32_to_cpu(hw_config->tdm_slots)) {
+			current_config = le32_to_cpu(hw_config->id);
+			matching_params = 3;
+			/* best match found */
 			break;
+		} else if (matching_params < 2 &&
+			   params_rate(params) == le32_to_cpu(hw_config->fsync_rate) &&
```

Comment:
Change second level to rate and channels and remove third level? Fail if first or second level match is not found. Copier can convert sample format but not rate and channels count.

Reply:
let me think about it a bit, this function was not able to fail before, even if it has not found the config.

Reply:
There could of course be some simple ranking about the formats if exact match is not found to prefer higher precision over lower, but in practice (only having s32 format available) this hardly matters.

Reply:
I think with this we will find the exact match and if not exact config available then we pick the first one which have matching rate and channels. It can be up to debate which non exact match is the best one, but I feel that if we don't have the exact match available then we are already in a 'this is not going to end up right' space..